File: frontend/src/lib/litsearch.ts
import { OpenAPI } from "@/client"

// backend/app/api/routes/litsearch.py + backend/app/schemas/litsearch.py:
// these three endpoints are new (Task 12) and not covered by the generated
// OpenAPI client, so — mirroring postChatMessage.ts — we hand-roll an auth'd
// fetch and hand-declare the response shapes here.

export type LitPaperFetchStatus =
  | "pending"
  | "downloading"
  | "done"
  | "failed"
  | "skipped"

export type LitPaperFulltextStatus = "none" | "added" | "failed"

export type LitPaperIngestStatus =
  | "none"
  | "queued"
  | "running"
  | "done"
  | "failed"

export type LiteratureSearchStage =
  | "searching"
  | "fetching"
  | "reading"
  | "done"
  | "failed"

export type LitAnswerKind = "abstracts" | "fulltext"

export interface LitAnswerRef {
  message_id: string
  kind: LitAnswerKind
}

export interface LiteraturePaperPublic {
  id: string
  doi: string | null
  title: string
  authors: string
  year: number | null
  abstract: string
  pdf_url: string | null
  citation_count: number | null
  fetch_status: LitPaperFetchStatus
  fulltext_status: LitPaperFulltextStatus
  fulltext_chars: number
  ingest_status: LitPaperIngestStatus
  document_id: string | null
}

export interface LiteratureSearchPublic {
  id: string
  stage: LiteratureSearchStage
  round: number
  followup_search_id: string | null
  papers: LiteraturePaperPublic[]
  answers: LitAnswerRef[]
  queries: string[]
}

export interface PaperIngestStatusPublic {
  status: string
  progress: number
  stage_name: string | null
  error: string | null
}

function authHeaders(): Record<string, string> {
  const token = localStorage.getItem("access_token")
  return token ? { Authorization: `Bearer ${token}` } : {}
}

export async function getSearch(
  searchId: string,
): Promise<LiteratureSearchPublic> {
  const response = await fetch(`${OpenAPI.BASE}/api/v1/litsearch/${searchId}`, {
    method: "GET",
    headers: { ...authHeaders() },
  })

  if (!response.ok) {
    throw new Error(`Litsearch request failed with status ${response.status}`)
  }

  return (await response.json()) as LiteratureSearchPublic
}

export async function addToDatabase(
  paperId: string,
): Promise<LiteraturePaperPublic> {
  const response = await fetch(
    `${OpenAPI.BASE}/api/v1/litsearch/papers/${paperId}/add-to-database`,
    {
      method: "POST",
      headers: { ...authHeaders() },
    },
  )

  if (!response.ok) {
    throw new Error(
      `Add-to-database request failed with status ${response.status}`,
    )
  }

  return (await response.json()) as LiteraturePaperPublic
}

export async function getIngestStatus(
  paperId: string,
): Promise<PaperIngestStatusPublic> {
  const response = await fetch(
    `${OpenAPI.BASE}/api/v1/litsearch/papers/${paperId}/ingest-status`,
    {
      method: "GET",
      headers: { ...authHeaders() },
    },
  )

  if (!response.ok) {
    throw new Error(
      `Ingest-status request failed with status ${response.status}`,
    )
  }

  return (await response.json()) as PaperIngestStatusPublic
}

File: backend/app/schemas/litsearch.py
import uuid
from typing import Literal

from sqlmodel import Field, SQLModel

from app.models.litsearch import FetchStatus, FulltextStatus, LitIngestStatus, LitStage

__all__ = [
    "LiteratureRef",
    "LiteraturePaperPublic",
    "LitAnswerRef",
    "LiteratureSearchPublic",
    "PaperIngestStatusPublic",
]


# Компактная ссылка на literature search, встраивается в ChatMessageResponse.literature
# (litsearch -> chat integration, companion spec §4.5), mirrors SubgraphResponse opt-in slot.
class LiteratureRef(SQLModel):
    search_id: uuid.UUID
    paper_count: int


# Публичное представление LiteraturePaper (app.models.litsearch), companion spec §4.5.
class LiteraturePaperPublic(SQLModel):
    id: uuid.UUID
    doi: str | None = None
    title: str
    authors: str
    year: int | None = None
    abstract: str
    pdf_url: str | None = None
    citation_count: int | None = None
    fetch_status: FetchStatus
    fulltext_status: FulltextStatus
    fulltext_chars: int
    ingest_status: LitIngestStatus
    document_id: uuid.UUID | None = None


# Ссылка на chat-сообщение, отвечавшее на основе abstracts/fulltext литературы
# (companion spec §4.5).
class LitAnswerRef(SQLModel):
    message_id: uuid.UUID
    kind: Literal["abstracts", "fulltext"]


# Публичное представление LiteratureSearch (app.models.litsearch), companion spec §4.5.
class LiteratureSearchPublic(SQLModel):
    id: uuid.UUID
    stage: LitStage
    round: int
    followup_search_id: uuid.UUID | None = None
    papers: list[LiteraturePaperPublic] = Field(default_factory=list)
    answers: list[LitAnswerRef] = Field(default_factory=list)
    # Every search's question in this chat turn (anchor first), Task 4: the
    # panel used to only ever see the anchor's own question — this surfaces
    # the full turn-union so the user sees every query the model ran.
    queries: list[str] = Field(default_factory=list)


# Статус фонового ingest-таска для одного paper. Отдельно от IngestUploadResponse:
# та схема требует непустой task_id и IngestStatus без варианта "none", а здесь
# нужно уметь выразить "задачи ещё не было" (все поля конструируются по умолчанию).
class PaperIngestStatusPublic(SQLModel):
    status: str = "none"
    progress: float = Field(default=0.0, ge=0, le=1)
    stage_name: str | None = None
    error: str | None = None
